zebROS_ws/src/behaviors/src/path_to_apriltag.cpp

```diff
@@ -0,0 +1,251 @@
+        gmts.child_frame_id = "base_link";
+        gmts.transform = gmt;
+
+        // get offset, transform by inverse transform
+        geometry_msgs::Pose p;
+        tf2::doTransform(offset, p, gmts);
+        trajectory_msgs::JointTrajectoryPoint pt = generateTrajectoryPoint(p.position.x, p.position.y, getYaw(p.orientation));
+        return std::optional<trajectory_msgs::JointTrajectoryPoint>{pt};
+    }
+
+
+    void callback(const field_obj::DetectionConstPtr& msg) {
+        // need to use transforms...
+        // oh wait no we don't! we can tell spline srv what frame id it is relevant to
+        latest_ = *msg;
+    }
+
+    void imuCallback(const sensor_msgs::ImuConstPtr& msg) {
+        sensor_msgs::Imu msgNonPtr = *msg;
+        latestImuZ_ = getYaw(msgNonPtr.orientation);
+    }
+
+    void executeCB(const behavior_actions::PathToAprilTagGoalConstPtr &goal)
+    {
+        uint32_t id = goal->id;
+        double tagRot = goal->tagRotation;
+        auto pt_ = applyOffset(id, goal->offset, tagRot);
+
+        if (!pt_.has_value()) {
+            as_.setAborted(result_);
+            return;
+        }
+        auto pt = pt_.value();
+
+        base_trajectory_msgs::GenerateSpline spline_gen_srv;
+
+        spline_gen_srv.request.header = latest_.header; // timestamp
+        spline_gen_srv.request.header.frame_id = "base_link"; // correct frame id
+
+        // if there is an offset, get atan2(y, x), add angle between robot and apriltag
+
+        spline_gen_srv.request.points.resize(2);
+        spline_gen_srv.request.point_frame_id.resize(2);
+        spline_gen_srv.request.points[0] = generateTrajectoryPoint(0, 0, 0);
+        //ROS_INFO_STREAM(position.x << "," << position.y << " " << zRot << " " << latestImuZ << " " << zRot - latestImuZ);
+        // angle now relative to apriltag
+        spline_gen_srv.request.points[1] = pt;
+        spline_gen_srv.request.point_frame_id[0] = "base_link";
+        spline_gen_srv.request.point_frame_id[1] = "base_link";
+        base_trajectory_msgs::PathOffsetLimit path_offset_limit;
+        spline_gen_srv.request.path_offset_limit.push_back(path_offset_limit);
+
+        if (client_.call(spline_gen_srv))
+        {
+            ROS_INFO_STREAM("Called");
+        }
+        else
+        {
+            ROS_ERROR("Failed to call service");
+            return;
+        }
+
+        nav_msgs::Path path = spline_gen_srv.response.path;
+
+        ROS_INFO("Waiting for action server to start.");
+        // wait for the action server to start
+        ac_.waitForServer(); //will wait for infinite time
+
+        ROS_INFO("Action server started, sending goal.");
+        // send a goal to the action
+        path_follower_msgs::PathGoal pathGoal;
+        pathGoal.path = path;
+        pathGoal.waypointsIdx = spline_gen_srv.response.waypointsIdx;
+        pathGoal.waypoints = spline_gen_srv.response.waypoints;
+        ac_.sendGoal(pathGoal);
+
+        ros::Rate r(100);
```

Comment:
Again, a slower rate would be better for CPU use

Reply:
i'll set it to 30hz also, let me know if that's still too fast